### ISSUE TITLE
attribute errors when trying pytorch_lightning_simple example

### ISSUE THREAD

**Issue by Jimmy2027** [optuna/optuna#1186]
I have tried to run the code from the [pytorch_lightning_simple](https://github.com/optuna/optuna/blob/master/examples/pytorch_lightning_simple.py) example for my own model, but get attribute errors with the `LightningModule`.

## Expected behavior

I expect the code from the example to run without errors.
## Environment

- Optuna version: 1.3.0
- Python version: 3.7
- OS: macOS Catalina Version 10.15.3 
- (Optional) Other libraries and their versions: pytorch_lightning : 0.7.5

## Error messages, stack traces, or logs
```
[W 2020-04-29 14:29:15,613] Setting status of trial#0 as TrialState.FAIL because of the following error: TypeError("Can't instantiate abstract class LightningNet with abstract methods forward")
Traceback (most recent call last):
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 677, in _run_trial
    result = func(trial)
  File "/Users/Hendrik/Desktop/larry1/min_working.py", line 172, in objective
    model = LightningNet(train_dataset, val_dataset, trial)
TypeError: Can't instantiate abstract class LightningNet with abstract methods forward
Traceback (most recent call last):
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 331, in optimize
    func, n_trials, timeout, catch, callbacks, gc_after_trial, None
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 626, in _optimize_sequential
    self._run_trial_and_callbacks(func, catch, callbacks, gc_after_trial)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 656, in _run_trial_and_callbacks
    trial = self._run_trial(func, catch, gc_after_trial)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 677, in _run_trial
    result = func(trial)
  File "/Users/Hendrik/Desktop/larry1/min_working.py", line 172, in objective
    model = LightningNet(train_dataset, val_dataset, trial)
TypeError: Can't instantiate abstract class LightningNet with abstract methods forward

```
## Steps to reproduce
Run the minimal working example to reproduce the error.
## Reproducible examples (optional)

```python
import os

import numpy as np
import optuna
import pkg_resources
import pytorch_lightning as pl
import torch
from optuna.integration import PyTorchLightningPruningCallback
from pytorch_lightning.logging import LightningLoggerBase
from torch.utils.data import DataLoader, Dataset


class MyDataset(Dataset):
    def __init__(self, data, labels):
        super(MyDataset, self).__init__()
        self.data = data
        self.labels = labels

    def __len__(self):
        return len(self.data)

    def __getitem__(self, item):
        return torch.from_numpy(self.data[item][None, ...]).float(), torch.from_numpy(self.labels[item]).float()

x_train = x_val = [np.ones((12,35)) for i in range(5)]
y_train = y_val = [np.ones((1,11)) for i in range(5)]

train_dataset = MyDataset(x_train, y_train)
val_dataset = MyDataset(x_val, y_val)

if pkg_resources.parse_version(pl.__version__) < pkg_resources.parse_version("0.6.0"):
    raise RuntimeError("PyTorch Lightning>=0.6.0 is required for this example.")

class DictLogger(LightningLoggerBase):
    """PyTorch Lightning `dict` logger."""

    def __init__(self, version):
        super(DictLogger, self).__init__()
        self.metrics = []
        self._version = version

    def log_metrics(self, metrics, step=None):
        self.metrics.append(metrics)

    @property
    def version(self):
        return self._version

    @property
    def experiment(self):
        """Return the experiment object associated with this logger."""

    def log_hyperparams(self, params):
        """
        Record hyperparameters.
        Args:
            params: :class:`~argparse.Namespace` containing the hyperparameters
        """

    @property
    def name(self):
        """Return the experiment name."""
        return 'optuna'

class Flatten(torch.nn.Module):
    def __init__(self):
        super(Flatten, self).__init__()

    def forward(self, x):
        batch_size = x.shape[0]
        return x.view(batch_size, -1)

class Net1(torch.nn.Module):
    def __init__(self, trial):
        super(Net1, self).__init__()

        kernel_size1 = trial.suggest_int('kernel_size1', 2, 7)
        kernel_size2 = trial.suggest_int('kernel_size2', 1, 16)
        num_params = ((12 - kernel_size1 + 1) // 2) * ((35 - kernel_size2 + 1) // 1) * 64
        self.network = torch.nn.Sequential(
            torch.nn.Conv2d(1, 64, kernel_size=(kernel_size1, kernel_size2), stride=1, padding=0),
            torch.nn.ReLU(inplace=True),
            torch.nn.MaxPool2d((2, 1)),
            Flatten(),
            torch.nn.Linear(num_params, 50),
            torch.nn.ReLU(inplace=True),
            torch.nn.Linear(50, 11),
            torch.nn.Sigmoid()
        )

    def forward(self, x):
        return self.network(x)

class LightningNet(pl.LightningModule):
    def __init__(self, train_dataset, val_dataset, trial):
        super(LightningNet, self).__init__()
        self.train_dataset = train_dataset
        self.val_dataset = val_dataset
        # Be careful not to overwrite `pl.LightningModule` attributes such as `self.model`.
        self._model = Net1(trial)
    def train_dataloader(self):
        return DataLoader(self.train_dataset, batch_size=2048, num_workers=16)

    def val_dataloader(self):
        return DataLoader(self.val_dataset, batch_size=2048, num_workers=16)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-4)

    def training_step(self, batch, batch_index):
        data, target = batch
        output = self.forward(data)
        loss = self.dice_loss(output, target)
        logs = {'train_loss': loss}
        return {'loss': loss, 'log': logs}

    def validation_step(self, batch, batch_index):
        data, target = batch
        output = self.forward(data)
        loss = self.dice_loss(output, target)
        return {'val_loss': loss}

    def validation_epoch_end(self, outputs):
        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
        tensorboard_logs = {'val_loss': avg_loss}
        self.val_loss = avg_loss
        return {'val_loss': avg_loss, 'log': tensorboard_logs}

    def dice_loss(self, input, target):
        smooth = 1.

        iflat = input.view(-1)
        tflat = target.view(-1)
        intersection = (iflat * tflat).sum()

        return 1 - ((2. * intersection + smooth) /
                    (iflat.sum() + tflat.sum() + smooth))

    # def forward(self, x):
    #     return self._model(x)
    #
    # def get(self, item):
    #     return self.__dict__[item]

if not os.path.exists('optuna_dir'):
    os.mkdir('optuna_dir')

EPOCHS = 100
DIR = 'optuna_dir'
MODEL_DIR = os.path.join(DIR, "result")


def objective(trial):
    # PyTorch Lightning will try to restore model parameters from previous trials if checkpoint
    # filenames match. Therefore, the filenames for each trial must be made unique.
    checkpoint_callback = pl.callbacks.ModelCheckpoint(
        os.path.join(MODEL_DIR, "trial_{}".format(trial.number)), monitor="val_loss"
    )

    # The default logger in PyTorch Lightning writes to event files to be consumed by
    # TensorBoard. We create a simple logger instead that holds the log in memory so that the
    # final accuracy can be obtained after optimization. When using the default logger, the
    # final accuracy could be stored in an attribute of the `Trainer` instead.
    logger = DictLogger(trial.number)

    trainer = pl.Trainer(
        logger=logger,
        checkpoint_callback=checkpoint_callback,
        max_epochs=EPOCHS,
        gpus=0 if torch.cuda.is_available() else None,
        early_stop_callback = PyTorchLightningPruningCallback(trial, monitor="val_loss"),
    )
    model = LightningNet(train_dataset, val_dataset, trial)
    trainer.fit(model)
    return logger.metrics[-1]["val_loss"]


pruner = optuna.pruners.MedianPruner()

study = optuna.create_study(direction="minimize", pruner=pruner)
study.optimize(objective, n_trials=100, timeout=600)

print("Number of finished trials: {}".format(len(study.trials)))

print("Best trial:")
trial = study.best_trial

print("  Value: {}".format(trial.value))

print("  Params: ")
for key, value in trial.params.items():
    print("    {}: {}".format(key, value))
```

## Additional context (optional)

Resolving the abstract class conflicts by adding a `forward` and a `get` function to the LightningNet class (commented out in the minimal working example) results in following errors:
```
[W 2020-04-29 14:35:40,296] Setting status of trial#0 as TrialState.FAIL because of the following error: TypeError("'<' not supported between instances of 'Trainer' and 'int'")
Traceback (most recent call last):
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 677, in _run_trial
    result = func(trial)
  File "/Users/Hendrik/Desktop/larry1/min_working.py", line 174, in objective
    trainer.fit(model)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 793, in fit
    self.run_pretrain_routine(model)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 913, in run_pretrain_routine
    self.train()
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 347, in train
    self.run_training_epoch()
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 453, in run_training_epoch
    self.call_early_stop_callback()
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 793, in call_early_stop_callback
    self.early_stop_callback.on_epoch_end(self, self.get_model())
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/integration/pytorch_lightning.py", line 61, in on_epoch_end
    self._trial.report(current_score, step=epoch)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/trial.py", line 597, in report
    if step < 0:
TypeError: '<' not supported between instances of 'Trainer' and 'int'
Traceback (most recent call last):
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 331, in optimize
    func, n_trials, timeout, catch, callbacks, gc_after_trial, None
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 626, in _optimize_sequential
    self._run_trial_and_callbacks(func, catch, callbacks, gc_after_trial)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 656, in _run_trial_and_callbacks
    trial = self._run_trial(func, catch, gc_after_trial)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/study.py", line 677, in _run_trial
    result = func(trial)
  File "/Users/Hendrik/Desktop/larry1/min_working.py", line 174, in objective
    trainer.fit(model)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 793, in fit
    self.run_pretrain_routine(model)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 913, in run_pretrain_routine
    self.train()
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 347, in train
    self.run_training_epoch()
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 453, in run_training_epoch
    self.call_early_stop_callback()
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 793, in call_early_stop_callback
    self.early_stop_callback.on_epoch_end(self, self.get_model())
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/integration/pytorch_lightning.py", line 61, in on_epoch_end
    self._trial.report(current_score, step=epoch)
  File "/Users/Hendrik/opt/anaconda3/lib/python3.7/site-packages/optuna/trial.py", line 597, in report
    if step < 0:
TypeError: '<' not supported between instances of 'Trainer' and 'int'

```

**Comment by hvy** [comment]
Sorry I believe it's a know bug but please correct me if I'm wrong. Is it fixed by this ongoing PR https://github.com/optuna/optuna/pull/1013 you think?

**Comment by festeh** [comment]
As the error `TypeError: Can't instantiate abstract class LightningNet with abstract methods forward` states you need to implement `forward` method in the `LightningNet`. Note that the current lightning example in the repo has this method implemented.

**Comment by festeh** [comment]
Something like this
```
def forward(self, data):
    return self._model(data)
```

**Comment by hvy** [comment]
Thanks @festeh for addressing some of the issues.

As for `TypeError: '<' not supported between instances of 'Trainer' and 'int'`, it seems to have been fixed by https://github.com/optuna/optuna/pull/1013/files#diff-37446c9777fc86723ebc01eb61ccb696R51-R61, which is included in the latest release 1.4.0 (or in the master branch). So if you could update your Optuna to the latest version with e.g. `pip install optuna -U`, this issue should be resolved.

Please let me close this issue for the above reason and since it hasn't seen any recent activity. Feel free to reopen if the problem still exists.
